statedb/statedb.go

```diff
@@ -1,35 +1,720 @@
+method.  These databases are auxiliary key-values that don't belong to the
+blockchain state, and thus any value in the NoState databases won't be
+reflected in the StateDB.Root.  One of the usecases for the NoState database is
+to store auxiliary mappings used to optimize the capacity usage of merkletrees
+used for zkSNARKS, where the number of levels is of critical matter.  For
+example, we may want to build a census tree of babyjubjub public keys that will
+be used to prove ownership of the public key via a SNARK in order to vote.  If
+we build the merkle tree using the public key as path, we will have an
+unbalanced tree which requires more levels than strictly necessary.  On the
+other hand, if we use a sequential index as path and set the value to the
+public key, we achieve maximum balancing reducing the number of tree levels.
+But then we can't easily query for the existence of a public key in the tree to
+generate a proof.  In such a case, we can store the mapping of public key to
+index in the NoState database.
+*/
 package statedb
 
-type StorageType string
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/vocdoni/arbo"
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/db/prefixeddb"
+	"go.vocdoni.io/dvote/tree"
+)
+
+func appendCopy(s []byte, v ...byte) []byte {
+	// Copy s to ensure we don't reuse the slice's backing array.
+	appended := make([]byte, 0, len(s)+len(v))
+	appended = append(appended, s...)
+	appended = append(appended, v...)
+	// Make sure the appended's capacity is fixed, to prevent future appends from sharing memory.
+	return appended[:len(appended):len(appended)]
+}
+
+var separator byte = '/'
+
+var (
+	subKeyTree    []byte = []byte("t")
+	subKeyMeta    []byte = []byte("m")
+	subKeyNoState []byte = []byte("n")
+	subKeySubTree []byte = []byte("s")
+)
 
-const (
-	StorageTypeDisk   StorageType = "disk"
-	StorageTypeMemory StorageType = "mem"
+var (
+	pathVersion   []byte = []byte("v")
+	keyCurVersion []byte = []byte("current")
 )
 
-type StateDB interface {
-	Init(storagePath string, storageType StorageType) error
-	Version() uint64
-	LoadVersion(int64) error // zero means last version, -1 is the previous to the last version
-	AddTree(name string) error
-	Tree(name string) StateTree
-	TreeWithRoot(root []byte) StateTree
-	ImmutableTree(name string) StateTree // a tree version that won't change
-	Commit() ([]byte, error)             // Returns New Hash
-	Rollback() error
-	KeyDiff(root1, root2 []byte) ([][]byte, error) // list of inserted keys on root2 that are not present in root1
-	Hash() []byte
-	Close() error
-}
-
-type StateTree interface {
+func uint32ToBytes(v uint32) []byte {
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, v)
+	return b
+}
+
+func bytesToUint32(b []byte) uint32 {
+	return binary.LittleEndian.Uint32(b)
+}
+
+// join two paths with a '/' in between
+func join(pathA, pathB []byte) []byte {
+	return appendCopy(appendCopy(pathA, separator), pathB...)
```

Comment:
you're calling `join` from many places, so it's unfortunate that it allocates multiple times. you should be able to allocate once, for `len(pathA)+len(pathB)+1`.

Reply:
also, a thought for you: if you used string instead of []byte for prefixes and separators/paths, then your code would get simpler. you'd use `path.Join` instead of `join`, and `appendCopy` would be unnecessary as Go strings are immutable so additions won't overwrite memory.

Reply:
I've replaced the exposed `[]byte` types by strings, and simplified a few things following your advice :)